api/src/main/java/io/opencensus/common/Timestamp.java

```diff
@@ -45,101 +46,116 @@ private Timestamp(long seconds, int nanos) {
-    } else {
-      return new Timestamp(seconds, nanos);
-    }
+  public static Timestamp fromMillis(long epochMilli) {
+    long secs = floorDiv(epochMilli, MILLIS_PER_SECOND);
+    int mos = (int) floorMod(epochMilli, MILLIS_PER_SECOND);
+    return create(secs, (int) (mos * NANOS_PER_MILLI)); // Safe int * NANOS_PER_MILLI
   }
 
   /**
    * Returns the number of seconds since the Unix Epoch represented by this timestamp.
    *
    * @return the number of seconds since the Unix Epoch.
    */
-  public long getSeconds() {
-    return seconds;
-  }
+  public abstract long getSeconds();
 
   /**
    * Returns the number of nanoseconds after the number of seconds since the Unix Epoch represented
    * by this timestamp.
    *
    * @return the number of nanoseconds after the number of seconds since the Unix Epoch.
    */
-  public int getNanos() {
-    return nanos;
-  }
+  public abstract int getNanos();
 
   /**
    * Returns a {@code Timestamp} calculated as this {@code Timestamp} plus some number of
    * nanoseconds.
    *
-   * @param nanos the nanoseconds to be added to the current timestamp.
-   * @return a {@code Timestamp} calculated as this {@code Timestamp} plus some number of
-   *     nanoseconds.
+   * @param nanosToAdd the nanos to add, positive or negative.
+   * @return the calculated {@code Timestamp}. For invalid inputs, a {@code Timestamp} of zero is
+   *     returned.
+   * @throws ArithmeticException if numeric overflow occurs.
    */
-  public Timestamp addNanos(long nanos) {
-    long newSeconds = seconds + nanos / NUM_NANOS_PER_SECOND;
-    nanos %= NUM_NANOS_PER_SECOND;
-    // Cannot overflow because: abs(nanos) < NUM_NANOS_PER_SECOND AND
-    // this.nanos < NUM_NANOS_PER_SECOND.
-    long newNanos = nanos + this.nanos;
-    newSeconds += (newNanos / NUM_NANOS_PER_SECOND);
-    newNanos %= NUM_NANOS_PER_SECOND;
-    if (newNanos >= 0) {
-      return Timestamp.create(newSeconds, (int) newNanos);
-    } else {
-      return Timestamp.create(newSeconds - 1, (int) (newNanos + NUM_NANOS_PER_SECOND));
-    }
+  public Timestamp addNanos(long nanosToAdd) {
+    return plus(0, nanosToAdd);
+  }
+
+  /**
+   * Returns a {@code Timestamp} calculated as this {@code Timestamp} plus some {@code Duration}.
+   *
+   * @param duration the {@code Duration} to add.
+   * @return a {@code Timestamp} with the specified {@code Duration} added.
+   */
+  public Timestamp addDuration(Duration duration) {
+    return plus(duration.getSeconds(), duration.getNanos());
   }
 
-  @Override
-  public boolean equals(Object obj) {
-    if (obj == this) {
-      return true;
+  /**
+   * Returns a {@link Duration} calculated as: {@code this - timestamp}.
+   *
+   * @param timestamp the {@code Timestamp} to subtract.
+   * @return the calculated {@code Duration}. For invalid inputs, {@link Duration#ZERO} is returned.
+   */
+  public Duration minusTimestamp(Timestamp timestamp) {
```

Comment:
How about subtractTimestamp?  I think it is more consistent with "add" in addDuration.